### PR TITLE
use property-decorators for no-member

### LINKED ISSUE #1127: False positive: no-member on "property"-like decorator
### Steps to reproduce
1. Have Pylint (1.6.4) evaluate the following snippet:
   
   ``` python
   class prop(object):
       def __init__(self, fn):
           self.value = 5
   
   class my_class(object):
       @prop
       def my_property(self):
           pass
   
   print(my_class().my_property.value)
   ```
### Current behavior

Pylint returns the following, even with `prop` added to `property_classes`:
    `E: 10, 6: Method 'my_property' has no 'value' member (no-member)`
### Expected behavior

No `no-member` errors. A decorated function should not be assumed to be a function.
### `pylint --version` output

```
    No config file found, using default configuration
    pylint 1.6.4, 
    astroid 1.4.8
    Python 3.5.2 (default, Jun 28 2016, 08:46:01) 
    [GCC 6.1.1 20160602]
```

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

`--property-classes=...` can now be configured to fix `no-member` FP.

Closes #1127

### CODE REVIEW

File: pylint/checkers/utils.py [hunk 1]
```diff
@@ -237,6 +238,8 @@
     {"_sitebuiltins.Quitter", "sys.exit", "posix._exit", "nt._exit"}
 )
 
+BUILTIN_PROPERTY = "builtins.property"
+
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:242]:
```suggestion
```

File: pylint/checkers/utils.py [hunk 2]
```diff
@@ -2190,3 +2193,19 @@ def is_terminating_func(node: nodes.Call) -> bool:
...
     return False
+
+
+def get_properties(config: argparse.Namespace) -> tuple[set[str], set[str]]:
+    """Returns a tuple of property classes and names.
+
+    Property classes are fully qualified, such as 'abc.abstractproperty' and
+    property names are the actual names, such as 'abstract_property'.
+    """
+    property_classes = {BUILTIN_PROPERTY}
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:2204]:
```suggestion
    property_classes = {"builtins.property"}
```
This is never going to change, we might as well make the code easier to read.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks great, I think this is a reasonable solution to a hard to fix problem. We need to document this better in the doc though because I forgot about it so chances are most users don't know about it either.

File: doc/user_guide/configuration/all-options.rst [hunk 3]
```diff
@@ -167,6 +167,13 @@ Standard Checkers
 **Default:**  ``True``
 
 
+--property-classes
+""""""""""""""""""
+*List of decorators that produce properties, such as abc.abstractproperty. Add to this list to register other decorators that produce valid properties. *
```
Comment (DanielNoord) [inline doc/user_guide/configuration/all-options.rst:172]:
```suggestion
*List of decorators that produce properties, such as abc.abstractproperty. Add to this list to register other decorators that produce valid properties.*
```

File: doc/whatsnew/fragments/1127.false_positive [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Fix ``no-name`` false positive in cases when property decorators are used.
```
Comment (DanielNoord) [inline doc/whatsnew/fragments/1127.false_positive:1]: It feels as if this entry should just be about the config. Any other issues should be fixed in a follow up.

File: pylint/lint/base_options.py [hunk 5]
```diff
@@ -401,6 +401,17 @@ def _make_linter_options(linter: PyLinter) -> Options:
                 "Useful if running pylint in a server-like mode.",
             },
         ),
+        (
+            "property-classes",
+            {
+                "default": ("abc.abstractproperty",),
```
Comment (DanielNoord) [inline pylint/lint/base_options.py:407]: This feels as if it should just be part of our own internal set.

File: pylint/lint/base_options.py [hunk 6]
```diff
@@ -401,6 +401,17 @@ def _make_linter_options(linter: PyLinter) -> Options:
...
+                "type": "csv",
+                "metavar": "<decorator names>",
+                "help": "List of decorators that produce properties, such as "
+                "abc.abstractproperty. Add to this list to register "
+                "other decorators that produce valid properties. ",
```
Comment (DanielNoord) [inline pylint/lint/base_options.py:412]:
```suggestion
                "other decorators that produce valid properties.",
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]

**Comment by clavedeluna** [comment]
> We need to document this better in the doc though because I forgot about it so chances are most users don't know about it either.

If you have suggestions for this PR please do share. Otherwise a separate issue would be good.

**Comment by Pierre-Sassoulas** [comment]
I was thinking of something similar to the doc for ``no-member`` where we explain in the message's doc what options can be used for false positives.

File: pylint/lint/base_options.py (hunk 5 above)
Comment (clavedeluna) [reply pylint/lint/base_options.py:407]: I think I feel otherwise as it gives maximum flexibility to a user. That's the end goal of config options.

Comment (DanielNoord) [reply pylint/lint/base_options.py:407]: Yeah, but then we should also remove `builtins.property`... This is just `stdlib`.

File: doc/data/messages/n/no-member/details.rst [hunk 7]
```diff
@@ -2,9 +2,10 @@ If you are getting the dreaded ``no-member`` error, there is a possibility that
 either:
 
 - pylint found a bug in your code
-- You're launching pylint without the dependencies installed in its environment
+- you're launching pylint without the dependencies installed in its environment
 - pylint would need to lint a C extension module and is refraining to do so
 - pylint does not understand dynamically generated code
+- your code uses a property decorator
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/n/no-member/details.rst:8]:
```suggestion
- your code uses a property decorator that pylint does not understand
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
LGTM !

File: pylint/lint/base_options.py (hunk 5 above)
Comment (DanielNoord) [reply pylint/lint/base_options.py:407]: Like I said. This should be hardcoded as this is just `stdlib`.

File: pylint/checkers/utils.py [hunk 8]
```diff
@@ -2192,6 +2193,22 @@ def is_terminating_func(node: nodes.Call) -> bool:
...
 
+def get_properties(config: argparse.Namespace) -> tuple[set[str], set[str]]:
+    """Returns a tuple of property classes and names.
+
+    Property classes are fully qualified, such as 'abc.abstractproperty' and
+    property names are the actual names, such as 'abstract_property'.
+    """
+    property_classes = {"builtins.property"}
+    property_names: set[str] = set()  # Not returning 'property', it has its own check.
+    if config is not None:
```
Comment (DanielNoord) [inline pylint/checkers/utils.py:2204]: This is unnecessary due to the typing.

File: pylint/checkers/base/name_checker/checker.py [hunk 9]
```diff
@@ -369,7 +340,6 @@ def visit_functiondef(self, node: nodes.FunctionDef) -> None:
                 if utils.has_known_bases(node.parent.frame(future=True))
                 else interfaces.INFERENCE_FAILURE
             )
```
Comment (DanielNoord) [inline pylint/checkers/base/name_checker/checker.py:342]:
```suggestion
            )

```

New lines help with readability.

File: pylint/checkers/base/name_checker/checker.py [hunk 10]
```diff
@@ -102,7 +85,7 @@ def _determine_function_name_type(
 
     :returns: One of ('function', 'method', 'attr')
     """
-    property_classes, property_names = _get_properties(config)
+    property_classes, property_names = utils.get_properties(config)
```
Comment (DanielNoord) [inline pylint/checkers/base/name_checker/checker.py:88]: This is not your PR, but it feels as if we should just use `property_classes` and always use inference.

File: tests/functional/n/no/no_member_property_decorator.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+class my_class:
+    @prop
+    def my_property(self):
+        pass
+
+print(my_class().my_property.value)
+
+
+def with_value(Cls):
+    Cls_setUp = Cls.setUp
```
Comment (DanielNoord) [inline tests/functional/n/no/no_member_property_decorator.py:23]: This feels really unpythonic. Can we make the tests a little more pythonic name- and code-wise?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
The code itself looks good and is a nice improvement. Just some nits on tests and my previous commetns

File: tests/functional/n/no/no_member_property_decorator.py (hunk 11 above)
Comment (clavedeluna) [reply tests/functional/n/no/no_member_property_decorator.py:23]: Sure, tho it's worth noting I copied it directly from the issue to demonstrate that it works.

**Comment by clavedeluna** [comment]
I _think_ the two pipeline failures aren't legitimate?

**Comment by clavedeluna** [comment]
I'm baffled by the 2 failures. pypy-38 I ran locally and it passes...

File: pylint/checkers/base/name_checker/checker.py (hunk 10 above)
Comment (DanielNoord) [reply pylint/checkers/base/name_checker/checker.py:93]: Are you going to address this issue?

Comment (clavedeluna) [reply pylint/checkers/base/name_checker/checker.py:93]: > This is not your PR

interpreted this as "not needed to do in this PR". I've received previous feedback that we should keep changes limited. Seeing how `get_properties` is used elsewhere it feels safer not to change htis at this time.

Comment (Pierre-Sassoulas) [reply pylint/checkers/base/name_checker/checker.py:93]: Yeah let's make it easy to merge PR we don't need to fix all old issues in the affected code to merge only to fix one issue :)

**Comment by clavedeluna** [comment]
Super close to being able to merge this PR but I keep trying to debug the readthedocs failure but I can't figure it out. If any maintainer could help that would be awesome

**Comment by Pierre-Sassoulas** [comment]
The error might be : ``/home/docs/checkouts/readthedocs.org/user_builds/pylint/checkouts/8002/doc/whatsnew/2/2.16/index.rst:45: ERROR: Unexpected indentation.`` Maybe undetected in 2.16.0 release ?

**Comment by clavedeluna** [comment]
> Details

oh gosh good eye. I grepped for error and there are just too many. let's see how I can reproduce and fix this 🤔

**Comment by clavedeluna** [comment]
Hm, small change to index.rst didn't fix it. I'm testing blindly here. any way to reproduce locally?

**Comment by clavedeluna** [comment]
@Pierre-Sassoulas any suggestions for this one? So close to merging, just stuck on the readthedocs failure that.I can't repro locally

**Comment by jacobtylerwalls** [comment]
In #9308 I just suggest bailing out of `no-member` when decorators are involved, just because our support for them is so poor. I wonder if we should put effort into understanding decorators better and then aggressively ignore messages in the meantime rather than adding config options we will need to maintain for a long time?

**Comment by Pierre-Sassoulas** [comment]
Seeing that this merge request is stuck, it's probably wiser.

**Comment by jacobtylerwalls** [comment]
Thanks for the investigation here. Closing in favor of a smaller-scoped fix in #9308.

**Pull request closed without merging**